File: app/src/main/java/by/d1makrat/library_fm/ui/fragment/top/TopItemsFragment.java
package by.d1makrat.library_fm.ui.fragment.top;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import by.d1makrat.library_fm.R;
import by.d1makrat.library_fm.asynctask.GetTopItemsCallback;
import by.d1makrat.library_fm.operation.model.TopOperationResult;
import by.d1makrat.library_fm.ui.fragment.ItemsFragment;

import static by.d1makrat.library_fm.Constants.PERIOD_KEY;

public abstract class TopItemsFragment<T> extends ItemsFragment<T> implements GetTopItemsCallback<T> {

    protected static final int MENU_SCROBBLES_OF_ARTIST = 0;
    protected static final int MENU_SCROBBLES_OF_TRACK = 1;
    protected static final int MENU_SCROBBLES_OF_ALBUM = 2;

    protected String mPeriod;
    protected TextView listHeadTextView;
    private String mTotalItemCount;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getArguments() != null) mPeriod = getArguments().getString(PERIOD_KEY);

        loadItems();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_refresh:
                if (!isLoading) {
                    allIsLoaded = false;

                    killTaskIfRunning(mGetItemsAsynctask);

                    mListAdapter.removeAll();
                    setUpListHead(mTotalItemCount, View.GONE);

                    mPage = 1;
                    loadItems();
                }
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.list_with_head, container, false);

        setUpRecyclerView(rootView);
        setUpActionBar((AppCompatActivity) getActivity());

        listHeadTextView = rootView.findViewById(R.id.list_head);

        if (mListAdapter.isEmpty()) {
            setUpListHead(mTotalItemCount, View.GONE);
        } else {
            setUpListHead(mTotalItemCount, View.VISIBLE);
        }

        return rootView;
    }

    @Override
    public void onLoadingSuccessful(TopOperationResult<T> result) {
        isLoading = false;

        mListAdapter.removeAllHeadersAndFooters();

        List<T> items = result.getItems();
        mTotalItemCount = result.getCount();

        int size = items.size();
        if (size > 0) {
            mListAdapter.addAll(items);
            setUpListHead(mTotalItemCount, View.VISIBLE);
        }
        else if (mListAdapter.isEmpty()){
            mListAdapter.addEmptyHeader(getString(R.string.no_items));
        }

        checkIfAllIsLoaded(size);
    }

    protected abstract void setUpListHead(String pItemsCount, int pVisibility);
}

File: app/src/main/java/by/d1makrat/library_fm/ui/fragment/top/TopItemsFragment.kt
package by.d1makrat.library_fm.ui.fragment.top

import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import by.d1makrat.library_fm.Constants
import by.d1makrat.library_fm.Constants.PERIOD_KEY
import by.d1makrat.library_fm.R
import by.d1makrat.library_fm.presenter.fragment.top.TopItemsPresenter
import by.d1makrat.library_fm.ui.CenteredToast
import by.d1makrat.library_fm.ui.fragment.ItemsFragment
import by.d1makrat.library_fm.view.fragment.TopItemsView
import java.text.DecimalFormat

abstract class TopItemsFragment<T>: ItemsFragment<T, TopItemsView<T>, TopItemsPresenter<T>>(), TopItemsView<T> {

    protected val MENU_SCROBBLES_OF_ARTIST = 0
    protected val MENU_SCROBBLES_OF_TRACK = 1
    protected val MENU_SCROBBLES_OF_ALBUM = 2

    protected var mPeriod: String? = null
    protected var listHeadTextView: TextView? = null

    protected lateinit var listHeadMessage: String
    protected lateinit var allIsLoadedMessage: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mPeriod = arguments?.getString(PERIOD_KEY)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.list_with_head, container, false)

        presenter?.attachView(this)

        setUpRecyclerView(rootView)

        listHeadTextView = rootView.findViewById(R.id.list_head)

        if (mListAdapter!!.isEmpty) {
            presenter?.onCreatingNewView()

        } else {
            presenter?.onShowingFromBackStack()
        }

        return rootView
    }

    override fun showEmptyHeader() {
        mListAdapter?.addEmptyHeader(getString(R.string.empty_rankings))
    }

    override fun getListItemsCount(): Int {
        return mListAdapter?.itemCount ?: 0
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        return when (item?.itemId) {
            R.id.action_refresh -> {
                presenter?.onRefresh()
                true
            }
            R.id.open_in_browser -> {
                presenter?.onOpenInBrowser()
                true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun hideListHead() {
        listHeadTextView?.visibility = View.INVISIBLE
    }

    override fun showListHead(itemCount: Int) {
        val formattedItemCount = DecimalFormat(Constants.NUMBER_FORMATTING_PATTERN).format(itemCount)
        listHeadTextView?.text = listHeadMessage.format(formattedItemCount)
        listHeadTextView?.visibility = View.VISIBLE
    }

    override fun showAllIsLoaded() {
        CenteredToast.show(context, allIsLoadedMessage, Toast.LENGTH_SHORT)
    }
}
